naming/dns_resolver.go

```diff
@@ -65,29 +64,34 @@ func formatIP(addr string) (addrIP string, ok bool) {
 // target: "www.google.com" returns host: "www.google.com", port: "443"
 // target: "ipv4-host:80" returns host: "ipv4-host", port: "80"
 // target: "[ipv6-host]" returns host: "ipv6-host", port: "443"
+// target: ":80" returns host: "localhost", port: "80"
+// target: ":" returns host: "localhost", port: "443"
 func setHostPort(target string) (host, port string, err error) {
 	if target == "" {
-		return "", "", errors.New(missingAddrErr)
+		return "", "", errMissingAddr
 	}
-	host, port = target, defaultPort
-	if _, ok := formatIP(target); !ok {
-		host, port, err = net.SplitHostPort(target)
-		if err != nil {
-			if err.(*net.AddrError).Err != missingPortErr {
-				return "", "", err
-			}
-			// The target is missing a port. We append the default port to the target.
-			host, port, err = net.SplitHostPort(target + ":" + defaultPort)
-			if err != nil {
-				return "", "", err
-			}
+
+	if ip := net.ParseIP(target); ip != nil {
+		// target is an IPv4 or IPv6(without brackets) address
+		return target, defaultPort, nil
+	}
+	if host, port, err := net.SplitHostPort(target); err == nil {
+		// target has port, i.e ipv4-host:port, [ipv6-host]:port, host-name:port
+		if host == "" {
+			// Keep consistent with net.Dial(): If the host is empty, as in ":80", the local system is assumed.
+			host = "localhost"
```

Comment:
It seems none of "localhost", "127.0.0.1", "[::1]", can work in all cases, we should probably just don't support empty name and return error in such case.